// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: aperture/tokenbucket/v1/state.proto

package tokenbucketv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on State with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *State) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on State with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in StateMultiError, or nil if none found.
func (m *State) ValidateAll() error {
	return m.validate(true)
}

func (m *State) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetStartFillAt()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, StateValidationError{
					field:  "StartFillAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, StateValidationError{
					field:  "StartFillAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetStartFillAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return StateValidationError{
				field:  "StartFillAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetLastFillAt()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, StateValidationError{
					field:  "LastFillAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, StateValidationError{
					field:  "LastFillAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetLastFillAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return StateValidationError{
				field:  "LastFillAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Available

	if len(errors) > 0 {
		return StateMultiError(errors)
	}

	return nil
}

// StateMultiError is an error wrapping multiple validation errors returned by
// State.ValidateAll() if the designated constraints aren't met.
type StateMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m StateMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m StateMultiError) AllErrors() []error { return m }

// StateValidationError is the validation error returned by State.Validate if
// the designated constraints aren't met.
type StateValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e StateValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e StateValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e StateValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e StateValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e StateValidationError) ErrorName() string { return "StateValidationError" }

// Error satisfies the builtin error interface
func (e StateValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sState.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = StateValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = StateValidationError{}

// Validate checks the field values on TakeNRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *TakeNRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TakeNRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in TakeNRequestMultiError, or
// nil if none found.
func (m *TakeNRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *TakeNRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetDeadline()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TakeNRequestValidationError{
					field:  "Deadline",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TakeNRequestValidationError{
					field:  "Deadline",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDeadline()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TakeNRequestValidationError{
				field:  "Deadline",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Want

	// no validation rules for CanWait

	if len(errors) > 0 {
		return TakeNRequestMultiError(errors)
	}

	return nil
}

// TakeNRequestMultiError is an error wrapping multiple validation errors
// returned by TakeNRequest.ValidateAll() if the designated constraints aren't met.
type TakeNRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TakeNRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TakeNRequestMultiError) AllErrors() []error { return m }

// TakeNRequestValidationError is the validation error returned by
// TakeNRequest.Validate if the designated constraints aren't met.
type TakeNRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TakeNRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TakeNRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TakeNRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TakeNRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TakeNRequestValidationError) ErrorName() string { return "TakeNRequestValidationError" }

// Error satisfies the builtin error interface
func (e TakeNRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTakeNRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TakeNRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TakeNRequestValidationError{}

// Validate checks the field values on TakeNResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *TakeNResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TakeNResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in TakeNResponseMultiError, or
// nil if none found.
func (m *TakeNResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *TakeNResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetAvailableAt()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TakeNResponseValidationError{
					field:  "AvailableAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TakeNResponseValidationError{
					field:  "AvailableAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetAvailableAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TakeNResponseValidationError{
				field:  "AvailableAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Current

	// no validation rules for Remaining

	// no validation rules for Ok

	if len(errors) > 0 {
		return TakeNResponseMultiError(errors)
	}

	return nil
}

// TakeNResponseMultiError is an error wrapping multiple validation errors
// returned by TakeNResponse.ValidateAll() if the designated constraints
// aren't met.
type TakeNResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TakeNResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TakeNResponseMultiError) AllErrors() []error { return m }

// TakeNResponseValidationError is the validation error returned by
// TakeNResponse.Validate if the designated constraints aren't met.
type TakeNResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TakeNResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TakeNResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TakeNResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TakeNResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TakeNResponseValidationError) ErrorName() string { return "TakeNResponseValidationError" }

// Error satisfies the builtin error interface
func (e TakeNResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTakeNResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TakeNResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TakeNResponseValidationError{}
